xml_converter/src/xml_converter.cpp

```diff
@@ -61,30 +63,115 @@ void write_xml_file(string xml_filepath, map<string, Category>* marker_categorie
+        if (pos != category_includes.end()) {
+            if (keep < i) {
+                proto_category->mutable_children()->SwapElements(i, keep);
+                if (proto_category->children(i).children_size() >= 0) {
+                    for (int j = 0; j < proto_category->children_size(); j++) {
+                        remove_proto_child(proto_category->mutable_children(j), category_includes);
+                    }
+                }
+            }
+            ++keep;
+        }
+    }
+
+    proto_category->mutable_children()->DeleteSubrange(keep, proto_category->children_size() - keep);
+}
+
 void write_protobuf_file(string proto_filepath, map<string, Category>* marker_categories, vector<Parseable*>* parsed_pois) {
-    ofstream outfile;
-    waypoint::Waypoint proto_message;
+    waypoint::Waypoint proto_pois;
+    waypoint::Waypoint all_categories;
+    waypoint::Waypoint output_message;
+    std::set<int> map_ids;
+    std::set<string> proto_categories;
+    std::set<string> category_includes;
+
+    if (mkdir(proto_filepath.c_str(), 0700) == -1) {
+        cout << "Error making ./protobins" << endl;
+        throw std::error_code();
+    }
 
-    outfile.open(proto_filepath, ios::out | ios::binary);
     for (const auto& category : *marker_categories) {
         waypoint::Category proto_category = category.second.as_protobuf();
-        proto_message.add_category()->CopyFrom(proto_category);
+        all_categories.add_category()->CopyFrom(proto_category);
+        create_proto_categories(proto_category, &proto_categories);
     }
 
     for (const auto& parsed_poi : *parsed_pois) {
         if (parsed_poi->classname() == "POI") {
             Icon* icon = dynamic_cast<Icon*>(parsed_poi);
             waypoint::Icon poi = icon->as_protobuf();
-            proto_message.add_icon()->CopyFrom(poi);
+            proto_pois.add_icon()->CopyFrom(poi);
+            map_ids.insert(icon->map_id);
         }
         if (parsed_poi->classname() == "Trail") {
             Trail* trail = dynamic_cast<Trail*>(parsed_poi);
             waypoint::Trail poi = trail->as_protobuf();
-            proto_message.add_trail()->CopyFrom(poi);
+            proto_pois.add_trail()->CopyFrom(poi);
+            map_ids.insert(trail->map_id);
         }
     }
-    proto_message.SerializeToOstream(&outfile);
-    outfile.close();
+
+    for (int map_id : map_ids) {
+        ofstream outfile;
+        string output_filepath = proto_filepath + "/" + to_string(map_id) + ".data";
+        outfile.open(output_filepath, ios::out | ios::binary);
+        output_message.CopyFrom(all_categories);
+
+        for (int i = 0; i < proto_pois.icon_size(); i++) {
+            vector<string> split_categories = split(proto_pois.icon(i).category().name(), ".");
+            if (proto_pois.icon(i).map_id() == map_id) {
+                auto category = proto_categories.find(split_categories[split_categories.size() - 1]);
```

Comment:
Will this return a false positive if there is a category such as
`mypack.hardpath.trail`

and you are looking for
`mypack.easypath.trail`

Reply:
Good point. I adjusted the set `category_includes` So it now captures parents names. From your example, the set would now include 

```
mypack
mypack.easypath
mypack.easypath.trail
```

Reply:
I reworded the functions so that the POIs now search for the entire name e.g. `mypack.hardpath.trail`

Whenever the name is used in the program it should now be in that form.